### ISSUE TITLE
Serializing anonymous objects throws an exception

### ISSUE THREAD

**Issue by NimaAra** [rpgmaker/NetJSON#126]
Given the following models:

```csharp
public struct LogEvent
{
	public DateTime Timestamp { get; set; }
	public Level Level { get; set; }
	public string Entry { get; set; }
}

public enum Level
{
	Debug,
	Trace
}
```
When trying to serialize the following:

```csharp
var logEvents = Enumerable.Range(1, 100)
		.Select(n => new LogEvent 
		{
			Timestamp = DateTime.UtcNow,
			Level = n % 2 == 0 ? Level.Debug : Level.Trace,
			Entry = n.ToString()
		});

var anonymousObjects = logEvents
    .Select(x => 
        new {
            TimestampEpoch = x.Timestamp,
            x.Level,
            Message = x.Entry
        }
    );

var defaultSettings = new NetJSONSettings { CamelCase = true, SkipDefaultValue = false };
var json = NetJSON.NetJSON.Serialize(anonymousObjects, defaultSettings);
```
Throws: 
_**TypeInitializationException: The type initializer for 'NetJSONCachedSerializer`1' threw an exception.**_

Removing the `defaultSettings` also results in the same error.

Note that serializing `logEvents` does not result in any exceptions and the problem is when the result is projected to _anonymous objects_.

**Comment by rpgmaker** [comment]
That is the expected behavior. I don't support non public types which includes auto-generated anonymous class. Supporting it requires a lot of effort since the code is only generating dynamic assemblies which do not let you make calls non public types.  The only thing that suppose that use is using dynamic methods which requires lot of rethinking on my part. 

Thanks,

**Comment by NimaAra** [comment]
Okay thanks for the quick reply.

This is unfortunate as in this example I had to do a projection and did not really want to create another POCO for the data. I just tried with `dynamic` and `ExpandoObject` with the same exception.

**Comment by rpgmaker** [comment]
That is understandable. The poco in this scenario would match your projection. Can't make promises on supporting it. If and when I do, it is going to be very slow in comparison to the current solution because it will need lots of reflection. I will browse the code and see if what it will be like to support it later tonight.

Thanks,

Thanks,

**Comment by NimaAra** [comment]
The POCO above is just an example as I kept it small for brevity in the scenario that I am trying to resolve the POCO does not match the projection.

I guess I can just create another POCO for now.

Thanks

**Comment by rpgmaker** [comment]
Cool. I will let you know after I review it tonight.

Thanks,

**Comment by rpgmaker** [comment]
All the changes needed will be quite a lot expecially been able to maintain both dynamic methods and assemblies in parallel due to non public types. I have figure a way to make them pay nicely together. I will try it and see if it works.

Thanks,

**Comment by oS-So** [comment]
Okay thanks, very much appreciated. 

I have created a secondary POCO for now but considering the inherent flexibility of JSON having support for serializing on-the-fly projections such as above would be a very nice feature to have.

**Comment by NimaAra** [comment]
Okay thanks, very much appreciated.

I have created a secondary POCO for now but considering the inherent flexibility of JSON having support for serializing on-the-fly projections such as above would be a very nice feature to have.

**Comment by rpgmaker** [comment]
Try the latest source. I added the initial code to support non-public type including anonymous types. Only serialize without settings option works at the moment. But can you try it and let me know if it works for you?

**Comment by wmjordan** [comment]
Congratulations! It could now serialize instances of internal or private classes.
I wrote an ugly class to test it.
The serialization result indicated that the outer class `C` could be serialized (with a public embedded class `P`), however it crashed when deserializing the string back to instance.
````C#
using System;
namespace ConsoleApplication1
{
	class Program
	{
		static void Main(string[] args) {
			C c = new C("a");
			string s;
			Console.WriteLine(s = c.Serialize());
			c = NetJSON.NetJSON.Deserialize<C>(s);
			Console.WriteLine(s = c.Serialize());
		}

		class C
		{
			public string S { get; set; }
			public int I { get; set; }
			public P Pdata { get; set; }
			D Data { get; set; }

			public class P
			{
				public int I { get; set; }
			}
			class D
			{
				public int A;
				public float B { get; set; }
				public C ClassC { get; set; }
			}

			public C() {

			}
			public C(string v) {
				S = v;
				I = Int32.MinValue;
				Pdata = new P { I = 123 };
				Data = new D {
					A = 1,
					B = (float)Math.PI,
					ClassC = new C() { I = Int32.MaxValue }
				};
			}
			public string Serialize() {
				return NetJSON.NetJSON.Serialize(this);
			}
		}
	}

}

````

**Comment by wmjordan** [comment]
The following case threw an `VerificationException`.
````C#
using System;
using System.Collections.Generic;
namespace ConsoleApplication1
{
	class Program
	{
		static void Main(string[] args) {
			string s;
			List<E> e = new List<E> { new E { V = 1 }, new E { V = 2 } };
			Console.WriteLine(s = NetJSON.NetJSON.Serialize(e));
			Console.WriteLine(NetJSON.NetJSON.Serialize(NetJSON.NetJSON.Deserialize<E>(s = NetJSON.NetJSON.Serialize(e))));
		}

		class E
		{
			public int V { get; set; }
		}
	}

}

````

**Comment by NimaAra** [comment]
@rpgmaker, you my friend is a wonder of nature!

The following worked like a charm:

```csharp
var json = NetJSON.Serialize(anonymousObjects);
var resultAsDynamic = NetJSON.Deserialize<dynamic>(json);
var resultAsObject = NetJSON.Deserialize<object>(json);
```

Though as you mentioned I cannot use the `NetJSONSettings` yet which means `SkipDefaultValue` is currently `true` which is not a big deal for now.

One other thing is that when trying to deserialize back to a compatible POCO e.g.

```csharp
public sealed class Projected
{
    public long Timestamp { get; set; }
    public Level Level { get; set; }
    public string Message { get; set; }
}
var resultAsProjected = NetJSON.Deserialize<Projected>(json);
```

It threw a `TypeLoadException`
```text
An exception of type 'System.TypeLoadException' occurred in NetJSONGeneratedAssembly but was not handled in user code

Additional information: Access is denied: 'NetJSON.NetJSONSerializer`1[System.Collections.Generic.IEnumerable`1[<>f__AnonymousType0`3[System.DateTime,NetJSON.Tests.Level,System.String]]]'.
```

I hope you did not have to sacrifice performance to get this working. FYI I am planning on putting together a performance regression test (if you don't have one already) at some point to keep track of any regressions.

Thank you.

**Comment by rpgmaker** [comment]
As I said before, I have only implemented one method to do the serialize so far for non public types. This was a test :). I am going to add the rest of the methods such as deserialization today. I only added the serialize for you to test if it was working properly using the way that I plan to support it.

Thanks,

**Comment by rpgmaker** [comment]
In regards to performance, non was sacrificed. I only redirect call from dynamic assemblies to dynamic methods to allow non public using the option that skip visibility check. Performance should not drop because I am still emitting same cil without use of reflection.

Thanks,

**Comment by rpgmaker** [comment]
A performance regression would be nice if you could do that since i don't have any at this time. I usually just run the ones that i found that other have created online such as burning monk benchmark.

Thanks,

**Comment by rpgmaker** [comment]
According to MSDN using dynamic method has less overhead than using dynamic assemblies so, it could be even faster who knows :).

https://msdn.microsoft.com/en-us/library/sfk2s47t%28v=vs.100%29.aspx?f=255&MSPPError=-2147217396

**Comment by NimaAra** [comment]
Okay, I hope I will have enough time to put together the performance regression tests this weekend. Last time I did some casual benchmarking it was even beating [Wire](https://github.com/akkadotnet/Wire). 

The problem with many of these benchmarking projects is that they either do not represent the proper payloads i.e. _small_, _medium_, _complex_, _large_ or they only concentrate on speed and not the GC pressure.

So coming up with a proper performance regression harness requires careful thinking.

**Comment by rpgmaker** [comment]
I looked at wire and it claim to be faster than everything lol. It even claim netserializer is the fastest binary whereas messageshark is faster than it. I don't create benchmarks, I usually let others to create it so it other people words for me nother mine.

**Comment by rpgmaker** [comment]
Can you give me the full source for the Projected class if possible? I have implemented the deserialize logic and want to test it locally first before i check it in

Thanks,

**Comment by oS-So** [comment]
Sure here's the full test.

```csharp
public struct LogEvent
{
    public DateTime Timestamp { get; set; }
    public Level Level { get; set; }
    public string Entry { get; set; }
}

public enum Level
{
    Debug,
    Trace
}

public sealed class Projected
{
    public long Timestamp { get; set; }
    public Level Level { get; set; }
    public string Message { get; set; }
}

[TestMethod]
public void When_serializing_anonymous_objects()
{
    var logEvents = Enumerable.Range(1, 100)
    .Select(n => new LogEvent
    {
        Timestamp = DateTime.UtcNow,
        Level = n % 2 == 0 ? Level.Debug : Level.Trace,
        Entry = n.ToString()
    });

    var anonymousObjects = logEvents
        .Select(x =>
            new
            {
                TimestampEpoch = x.Timestamp,
                x.Level,
                Message = x.Entry
            }
        );

    var json = NetJSON.Serialize(anonymousObjects);
    var resultAsDynamic = NetJSON.Deserialize<dynamic>(json);
    var resultAsObject = NetJSON.Deserialize<object>(json);
    var resultAsProjected = NetJSON.Deserialize<Projected>(json);
}
```

**Comment by NimaAra** [comment]
Sure here's the full test.

```csharp
public struct LogEvent
{
    public DateTime Timestamp { get; set; }
    public Level Level { get; set; }
    public string Entry { get; set; }
}

public enum Level
{
    Debug,
    Trace
}

public sealed class Projected
{
    public long Timestamp { get; set; }
    public Level Level { get; set; }
    public string Message { get; set; }
}

[TestMethod]
public void When_serializing_anonymous_objects()
{
    var logEvents = Enumerable.Range(1, 100)
    .Select(n => new LogEvent
    {
        Timestamp = DateTime.UtcNow,
        Level = n % 2 == 0 ? Level.Debug : Level.Trace,
        Entry = n.ToString()
    });

    var anonymousObjects = logEvents
        .Select(x =>
            new
            {
                TimestampEpoch = x.Timestamp,
                x.Level,
                Message = x.Entry
            }
        );

    var json = NetJSON.Serialize(anonymousObjects);
    var resultAsDynamic = NetJSON.Deserialize<dynamic>(json);
    var resultAsObject = NetJSON.Deserialize<object>(json);
    var resultAsProjected = NetJSON.Deserialize<Projected>(json);
}
```

**Comment by NimaAra** [comment]
@rpgmaker FYI, I have added NetJSON to the Wire performance tests and it has been merged into the dev branch.

You can run the [app](https://github.com/akkadotnet/Wire/tree/dev/Wire.PerfTest) to see the result, at the moment it is falling behind particularly on the deserialization.

**Comment by rpgmaker** [comment]
Can you post the result here? The difference in performance is because I don't have a custom handler for guidance. Wire is binary format, right?

**Comment by NimaAra** [comment]
Yes it is binary.

I will open another issue and post the result there so that it can be in it's own thread. FYI, `MessageShark` is also there though it is currently broken.

**Comment by rpgmaker** [comment]
Broken? How?

**Comment by NimaAra** [comment]
It's returning 0 bytes for some of the tests I haven't debugged it.

**Comment by rpgmaker** [comment]
That is expected. You need to enable include field option when serializing fields to not get zero bytes.

Thanks,

**Comment by NimaAra** [comment]
Added the result to https://github.com/rpgmaker/NetJSON/issues/127

**Comment by rpgmaker** [comment]
I meant to say use this version that supports fields: https://github.com/rpgmaker/MessageShark/commit/0ef6e788676b7fc52f169073752bb709048a329a 

Or simply convert the fields in test class to Getter/Setter
Thanks,

**Comment by NimaAra** [comment]
Is that version available on NuGet?

**Comment by rpgmaker** [comment]
I never made it available because it was a lot of code changes.

Thanks,

**Comment by rpgmaker** [comment]
Please test again. Issue has been resolved. Thanks for forcing me to think about how to implement anonymous and non public type serialization :)

**Comment by rpgmaker** [comment]
Next task is to improve performance to be closer or better than Wire serializer even though it is binary :)

**Comment by wmjordan** [comment]
@rpgmaker I examined the source code of `DynamicNetJSONSerializer` and found that it internally uses `ConcurrentDictionary` to look up serializers. Each time a serialization method or deserialization method is called, the dictionary must be looked up once and consequently performance can get hurt.
Why not eliminate the dictionary and bake all delegates in the constructor since the `DynamicNetJSONSerializer` is created once and then stored in `NetJSONCachedSerializer`?

**Comment by rpgmaker** [comment]
You are correct. I could eliminate the look up completely and access all the delegates as class instance members which will improve the performance when using non public types.I will make the change in my next checkin.

Thanks,

**Comment by NimaAra** [comment]
Hi,

That's great however there's a slight problem when deserializing to the `Projected` type.

The model should actually be: (forgot to change the `long` to `DateTime`):

```csharp
public sealed class Projected
{
    public DateTime Timestamp { get; set; }
    public Level Level { get; set; }
    public string Message { get; set; }
}
```

All the timestamps are deserialized to the default value of `DateTime` for some reason. Both `Dynamic` and `Object` are fine though.

The joy of working with `DateTime`! ;-)

**Comment by rpgmaker** [comment]
It is not the problem of datetime. You returned TimestampEpoch in your anonymous projection. You need to rename the property to Timestamp without "Epoch" on it. And also make sure that in your setting that you use Date format option to tell it that is Epoch based.

Thanks,

**Comment by NimaAra** [comment]
School boy error! works great thanks!

Any plan on releasing a NuGet? or are you waiting to make the perf improvements first?

**Comment by rpgmaker** [comment]
I will put it on nuget tonight. I will work on the performance stuff as needed

Thanks,

**Comment by rpgmaker** [comment]
Nuget package created https://www.nuget.org/packages/NetJSON/
